Remove stream editor email duplication

meadmaker opened the pull request:
These changes will prevent the stream editor email address from being included in the `Notice emails` field during the creation of a conflict review.  This only works for documents in the IRTF or ISE streams.
- If the stream editor email address is detected in the input, then the address is silently removed.  
- The address is removed even if it has been entered multiple times.
- The email address suggestion generator has been updated to remove the stream editors
- The help text for the `Notice emails` field has been updated to say that the stream editors are automatically notified.

While I was in the code, I refactored a bit:
- The SimpleStartReviewForm is now a superclass to the StartReviewForm.  This was intended to ease the creation of a single `clean_notify()` validation method, which eventually moved to the view code instead.  However, the refactoring was left.
- The `start_review_as_secretariat()` and `start_review_as_stream_owner()` methods are now merged.  Ironically, this resulted in the exact same number of lines of code, but at least there is only one method to update going forward.

Fixes ietf-tools/datatracker#3638.

meadmaker committed:
Add test for desired behavior

meadmaker committed:
Make StartReviewForm a subclass of SimpleStartReviewForm

When I make changes to the notify field, now I only have to change it in one place.

meadmaker committed:
Consolidate nearly identical methods

The start_review_as_secretariat and start_review_as_stream_owner methods were virtually identical.  Consolidate them using a few conditional checks.  Now the logic isn't duplicated, so we won't need to perform maintenance in two places.

meadmaker committed:
Remove stream managers from the notify list

This will:
- Find the stream managers by calling Recipient.gather()
- extract just their email addresses
- Split the 'notify' field by commas
- Remove any item in the notifications list that matches any of the stream manager emails.

Yes, this will incorrectly remove email addresses like:
  "rfc-ise@rfc-editor.org"<other_addy@example.com>
Anyone who enters that gets what they deserve.

meadmaker committed:
Move the address removal from the form to the view

Because the document name isn't available to the form, so I can't retrieve the document in the form's `clean()` method, so I can't discover the correct document stream in the form.

meadmaker committed:
Tell users that the stream editor is included automatically.

meadmaker committed:
Don't suggest that we include the stream editors just to be removed after submission.

meadmaker committed:
Verify an initial condition of doc stream being in IRTF or ISE

I expect that this will always be satisfied, but it's a good assumption validation.

jennifer-richards commented on ietf/doc/tests_conflict_review.py:
```diff
@@ -80,6 +81,63 @@ def test_start_review_as_secretary(self):
+        ad_strpk = str(Person.objects.get(name='Areað Irector').pk)
+        state_strpk = str(State.objects.get(used=True, slug='needshep', type__slug='conflrev').pk)
+        self.client.login(username="secretary", password="secretary+password")
+        """ This set of notification recipients should cover a bunch of cases for this test.
```
This should really be a comment - triple quotes aren't meant for multiline comments, notwithstanding their use as a docstring after a function def.

jennifer-richards commented on ietf/doc/tests_conflict_review.py:
```diff
@@ -80,6 +81,63 @@ def test_start_review_as_secretary(self):
+            
+            # Get the email addresses which should remain
+            other_emails = [manager_emails[stream] 
+                            for stream in list(manager_emails) 
```
Pythonic preference would be `[mgr_email for stream, mgr_email in manage_emails.items() if..]`. If you prefr as-is, you can drop the `list`. (A dict acts as an iterator over its keys.)

jennifer-richards commented on ietf/doc/views_conflict_review.py:
```diff
@@ -397,10 +397,35 @@ def __init__(self, *args, **kwargs):
 def start_review(request, name):
-    if has_role(request.user,"Secretariat"):
-        return start_review_as_secretariat(request,name)
+    """Start the conflict review process, setting the initial 
```
Docstring format should ideally be:
```python
def somefunc():
    '''Do the thing this method does in a single line

    After a blank line, describe in more detail / document parameters, etc.
    '''
```

jennifer-richards commented on ietf/doc/views_conflict_review.py:
```diff
@@ -375,6 +377,23 @@ class SimpleStartReviewForm(forms.Form):
+                
+        notify = cleaned_data['notify']
+        stream_managers = [
+            re.sub('([^<]*<)?(.*)(>.*)', '\2', r) 
```
Could use `email.utils.parseaddr()` here - it'll return a tuple, `(realname, email)`. If you apply that to addresses in the notifications list when filtering, I think that'd get rid of the caveat (although I 100% agree they get what they deserve if you choose to keep it simple)

jennifer-richards commented on ietf/doc/views_conflict_review.py:
```diff
@@ -455,13 +455,19 @@ def start_review_sanity_check(request, name):
+    if doc_to_review.stream.slug not in ['ise', 'irtf']:
+        notify_addresses.extend(
+            [r.formatted_email() 
+             for r in Role.objects.filter(group__acronym=doc_to_review.stream.slug, name='chair')]
```
Does this match the addresses now being filtered out from the notify list? Here, it's addresses for 'chair' Roles. The filtered items are 'stream_managers' from Recipients. Do these match up in the right way?

jennifer-richards reviewed (changes requested):
I have a bunch of nit picky suggestions plus one real question.

meadmaker committed:
Accept stylistic code review corrections

meadmaker committed:
Explicitly remove the addresses to avoid

Instead of avoiding adding document stream chair addresses as a currently accurate proxy for the stream editor addresses, use the gather_stream_managers() function to find the exact addresses to avoid and remove those from any address we consider.

meadmaker commented:
Back at ya!

meadmaker committed:
Merge branch 'ietf-tools:main' into mark/remove_stream_editor_email_duplication

Ebrahim4600 commented:
Good
